### PR TITLE
object_vectors is a list of lists

### PR DESCRIPTION
**Pull request by hobson**
* describe_state -> describe_scene  which accepts the new format list of object_vectors
* update_state_dict -> update_state  has no state and just uses estimate_color and object label to compose state vector (location from Ashwin TBD)
* estimate_color used zip to simplify

### CODE REVIEW

File: nlp/__init__.py [hunk 1]
```diff
@@ -1,2 +1 @@
-from nlp.core import update_state, update_state_dict, describe_state, say
-
+from nlp.core import update_state, describe_scene, say
```
Comment (alxrsngrtn) [inline nlp/__init__.py]: Did you rename `update_state_dict` to `update_state`? Otherwise, we need the old import.

File: nlp/core.py [hunk 2]
```diff
@@ -3,7 +3,8 @@
 import os
 
 import pandas as pd
-import object_detection.color_labeler as color_labeler
+# fix the antipattern of having a separate folder for every function/class
```
Comment (alxrsngrtn) [inline nlp/core.py]: Will do!

Comment (alxrsngrtn) [reply nlp/core.py]: ... eventually

File: nlp/core.py [hunk 3]
```diff
@@ -44,135 +45,58 @@ def pluralize(s):
...
-        }
+        list: list of object vectors, for example:
+        LABEL_KEYS = '.split()
+            [
+                ['cup', 0, .95, -.5, .1, 0, .1, .1, 0, .5, .3, .14, .01, .01, .01, .01, .01, .01]
+                ['ski', 0, .80, -.5, .1, 0, .1, .1, 0, .5, .3, .14, .01, .01, .01, .01, .01, .01]
+            ]
+            The object vector keys are defined in constants.OBJECT_VECTOR_KEYS:
+                [category instance confidence x y z width height depth
+                 black white red orange yellow green cyan blue purple pink]
```
Comment (alxrsngrtn) [inline nlp/core.py]: Using this defined constant, we can probably define a function to do a key-value lookup of the state vector.

File: nlp/__init__.py (hunk 1 above)
Comment (alxrsngrtn) [reply nlp/__init__.py]: I see that you did. This is preferred.

File: nlp/core.py [hunk 4]
```diff
@@ -44,135 +45,58 @@ def pluralize(s):
...
-            print(display_str)  # TODO(Alex) Convert to logging
-
-            obj_data = {
-                'score': scores[i],
-                'color': color_labeler.estimate(image, box=boxes[i])
-            }
-
-            state_obj[class_name] += [obj_data]
-
-    update_state_dict.i += 1
```
Comment (alxrsngrtn) [inline nlp/core.py]: Is this needed for the `say` command in the main script?

File: nlp/core.py [hunk 5]
```diff
@@ -44,135 +45,58 @@ def pluralize(s):
...
     num_boxes = min([boxes.shape[0] if max_boxes_to_draw is None else max_boxes_to_draw, boxes.shape[0], len(classes)])
 
-    state_obj = defaultdict(list)
-    if update_state_dict.i is None:
-       update_state_dict.i = 0
-
+    object_vectors = []
     for i in range(num_boxes):
         if scores is None or scores[i] > min_score_thresh:
+            # box = tuple(boxes[i].tolist())
```
Comment (alxrsngrtn) [inline nlp/core.py]: Remove

File: nlp/core.py [hunk 6]
```diff
@@ -44,135 +45,58 @@ def pluralize(s):
...
+    ...    ['cup', 0, .95, -.5, .1, 0, .1, .1, 0, .5, .3, .14, .01, .01, .01, .01, .01, .01]
+    ...    ['ski', 0, .80, -.5, .1, 0, .1, .1, 0, .5, .3, .14, .01, .01, .01, .01, .01, .01]
+    ... ]
+    >>> describe_scene(object_vectors)
     """
-    def count_objects(state_dict):
-        new_dict = {k: len(v) for k, v in state_dict.items()}
-        return new_dict
+    def count_objects(object_vectors):
+        return collections.Counter(list(zip(*object_vectors))[0]) if len(object_vectors) else {}
```
Comment (alxrsngrtn) [inline nlp/core.py]: Nice use of collections.

File: nlp/core.py [hunk 7]
```diff
@@ -44,135 +45,58 @@ def pluralize(s):
...
 
-    state_counts = count_objects(state)
+    object_counts = count_objects(object_vectors)
 
-    plural_description_list = ['{} {}'.format(i, pluralize(s) if i > 1 else s) for (s, i) in state_counts.items()]
+    plural_description_list = ['{} {}'.format(i, pluralize(s) if i > 1 else s) for (s, i) in object_counts.items()]
 
     comma_list = ', '.join(plural_description_list[:-2])
     conjunction = ' and '.join(plural_description_list[-2:])
```
Comment (alxrsngrtn) [inline nlp/core.py]: I have an update to the above two lines that I thought I pushed. I extracted them out as a function anyway, you'll see it soon.

File: object_detection/color_labeler/core.py [hunk 8]
```diff
@@ -16,40 +19,23 @@ def estimate_color(img, box=None):
...
+        white     0.3
+        red       0.5
+        orange    0.2
+        yellow    0.0
+        green     0.0
+        cyan      0.0
+        blue      0.0
+        purple    0.0
+        pink      0.0
+        Name: color, dtype: float64
```
Comment (alxrsngrtn) [inline object_detection/color_labeler/core.py]: Nicely formatted example! This looks like a good revision of the `estimate_color` function if this is the Input/Output.

Comment (alxrsngrtn) [reply object_detection/color_labeler/core.py]: Oh I see, you're flexing your pandas knowledge. Very nice.

File: object_detection/color_labeler/core.py [hunk 9]
```diff
@@ -93,23 +79,7 @@ def estimate_color(img, box=None):
...
-    od['blue'] = hist_hue[5] / total
-    od['purple'] = hist_hue[6] / total
-    od['pink'] = hist_hue[7] / total
-    output = pd.Series(od)
-
-    return output
-
-
-
+    return pd.Series([n_black, n_white] + list(hist_hue), index=COLOR_KEYS, name='color') / total
```
Comment (alxrsngrtn) [inline object_detection/color_labeler/core.py]: Ah, very elegant.

File: object_detection/constants.py [hunk 10]
```diff
@@ -20,6 +20,6 @@
 CATEGORY_INDEX = label_map_util.create_category_index(CATEGORIES)
 
 LABEL_KEYS = 'category instance'.split()
-COLOR_KEYS = 'red orange yellow green indigo violet black white pink'.split()
+COLOR_KEYS = 'black white red orange yellow green cyan blue purple pink'.split()
```
Comment (alxrsngrtn) [inline object_detection/constants.py]: should we change `cyan` to `light_blue` or something to that effect?

File: object_detection_app.py [hunk 11]
```diff
@@ -9,16 +9,15 @@
...
 from nlp.dispatch import mqttc, dispatcher
 from nlp.command import Describe, DescribeColor
 
 
-
 from object_detection.constants import CATEGORY_INDEX, PATH_TO_CKPT
 
 
-def detect_objects(image_np, sess, detection_graph, _state_q, utterance_frames=20, voice_on=False):
+def detect_objects(image_np, sess, detection_graph, _state_q, utterance_frames=1, voice_on=False):
```
Comment (alxrsngrtn) [inline object_detection_app.py]: Can you explain what exactly `utterance_frames` does?

File: object_detection_app.py [hunk 12]
```diff
@@ -48,17 +47,16 @@ def detect_objects(image_np, sess, detection_graph, _state_q, utterance_frames=2
...
-    _state_q.put(state)
+    # Persists image state (list of object vectors for that image) in a queue
+    _state_q.put(object_vectors)
 
-    if not update_state_dict.i % utterance_frames:
-        description = describe_state(state)
-        if voice_on:
-            say(description)
+    description = describe_scene(object_vectors)
+    if voice_on:
```
Comment (alxrsngrtn) [inline object_detection_app.py]: `utterance_frames` looks like it isn't being used.

File: object_detection_app.py (hunk 11 above)
Comment (alxrsngrtn) [reply object_detection_app.py]: After looking below, can we git rid of this argument?

File: nlp/core.py [hunk 13]
```diff
@@ -44,135 +45,58 @@ def pluralize(s):
...
-                [
-                    { color: [<color_vec>], score: 0.99, ... },
-                    { color: [<color_vec>], score: 0.75, ... },
-                    { color: [<color_vec>], score: 0.85, ... }
-                ],
-            ...
-
-        }
+        list: list of object vectors, for example:
+        LABEL_KEYS = '.split()
```
Comment (alxrsngrtn) [inline nlp/core.py]: I don't think `'.split()` is syntactically correct.

Comment (alxrsngrtn) [reply nlp/core.py]: Minor issue. Do you mean to say something like `_.split()`?

File: nlp/core.py (hunk 3 above)
Comment (alxrsngrtn) [reply nlp/core.py]: I'm a little worried about this design. It seems like if we make a change to the `OBJECT_VECTOR_KEYS`/order or number of keys in the vector, we'll have to make changes in many places of code. Is there a way we can encapsulate the object vector in an object to do this for us?

Comment (alxrsngrtn) [reply nlp/core.py]: This will certainly get us by for the NSF grant, but in the future, I think we shouldn't repeat this pattern. I know you've said you've used it before in the past in many projects, but I'd like a discussion with you about the tradeoffs.

File: object_detection/constants.py (hunk 10 above)
Comment (alxrsngrtn) [reply object_detection/constants.py]: or `lightblue`...

Comment (alxrsngrtn) [reply object_detection/constants.py]: Just tested, this works in OSX dictation.
